packages/dev/core/src/Materials/greasedLineSimpleMaterial.ts

```diff
@@ -0,0 +1,511 @@
+import type { Scene } from "../scene";
+import { Engine } from "../Engines/engine";
+import { RawTexture } from "./Textures/rawTexture";
+import type { GreasedLineMaterialOptions, IGreasedLineMaterial } from "./greasedLinePluginMaterial";
+import { GreasedLineMeshColorMode, GreasedLineMeshColorDistributionType } from "./greasedLinePluginMaterial";
+
+import { ShaderMaterial } from "./shaderMaterial";
+import type { Nullable } from "../types";
+import { Color3 } from "../Maths/math.color";
+import { TmpVectors, Vector2 } from "../Maths/math.vector";
+import type { BaseTexture } from "./Textures/baseTexture";
+
+import "../Shaders/greasedLine.fragment";
+import "../Shaders/greasedLine.vertex";
+
+/**
+ * GreasedLineSimpleMaterial
+ */
+export class GreasedLineSimpleMaterial extends ShaderMaterial implements IGreasedLineMaterial {

```

Comment:
Is it not possible to create a common base class for `GreasedLineSimpleMaterial` and `GreasedLinePluginMaterial`? 95% of the code is the same...

The main problem is that `GreasedLinePluginMaterial` already extends `MaterialPluginBase` and won't be able to extend another class...

I don't know if there's a way around that, cc @RaananW.

If there's no other solution, I think composition is still better than duplicating the code, because we would have all the code in a single location, and `GreasedLinePluginMaterial` / `GreasedLineSimpleMaterial` would simply defer the execution to the corresponding methods of the common class.

Reply:
With all the respect but how are these classes identical?

Reply:
I meant that `_BooleanToNumber`, `_Color3toRGBAUint8`, `_createColorsTexture`, `dispose`, `updateLazy`, `setColors`, `serialize`, `parse` and `_PrepareEmptyColorsTexture` are the same methods in both cases and with the same implementations, minus some minor adjustements for some of them that can be easily dealt with. And they are the methods that actually implement the greased line feature.

They also have the same properties. The difference is that the setters call the `setXXX` method to set the value in the `ShaderMaterial` case. I wonder if using a decorator would allow to add this additional bit of code in the `ShaderMaterial` case...

Reply:
The `public static` functions can be shared. That's true. But some version of webpack can't tree shake the correctly and the whole reffered class gets bundled not just the `public static` function from the class. I believe @RaananW is expert on this so he can shed more light on this issue.

IMHO there is no reason to make the code more complex, the resulting compiled JS code might be longer and slower. Those instance methods which are identicall are only a few lines long. 

Another reason to have them in both classes is that the `SimpleGreasedLineMaterial` will not be enhanced anymore, it already contains everything it needs. The plugin material could be updated with new functionality in the future and the instance methods will differ a lot more in this case.

Reply:
> Another reason to have them in both classes is that the SimpleGreasedLineMaterial will not be enhanced anymore, it already contains everything it needs. 

I'm not so sure! This is the main reason why I suggest factoring the code related specifically to the greased line feature, because if you (or anyone else) add new functionality to the plugin version, there's no reason why we shouldn't have the same functionality in the shading material version. It's hard to say today that we're feature complete and will never add new features in the future that could benefit both versions.

In any case, I'm fine if it's too much work or too cumbersome to extract this code into an external class!